### ISSUE TITLE
How does Katago solve the following problem: in unfavorable situations, fast search visits high-prior moves too little, resulting in very poor decisions?

### ISSUE THREAD

**Issue by Garbage123King** [lightvector/KataGo#1046]
Consider a fast-search MCTS with total_visits = 100, just like in the paper. Using the PUCT formula: PUCT = q + u, where in my implementation q is in the range [-1, 1].

When using the PUCT formula, if the position is unfavorable, q will be negative, reducing the PUCT value of the node to below zero. However, even the worst nodes still end up with a slightly positive PUCT value, because they got q=0 and u>0. As a result, MCTS starts heavily exploring completely hopeless nodes — for example, those with a prior of 0.0002. Then, around visit 50, due to the increase in u, MCTS finally explores a high-prior move once. After that, since the q is negative again, the PUCT drops below zero and the search again focuses on hopeless nodes.

In the end, the top candidate (prior = 0.39) was only visited 4 times, the second candidate (prior = 0.17) was visited just 3 times. The top candidate was actually very bad and led to a direct loss, but because it had a slight advantage in visits, it was ultimately chosen. The second candidate was actually much better, but due to the low number of visits, it never got a fair chance. (At the end of the search, the top candidate had a q of -0.96, while the second had a better q of -0.6.)

So, out of these 100 searches, fewer than 10 were actually meaningful. I tried increasing total_visits from 100 to 200, but the situation didn’t improve — which makes sense, since as long as there's space left, the PUCT values of those empty positions will remain above 0. I must be misunderstanding some key detail here, right?

**Comment by lightvector** [comment]
This is something you're observing in your own custom MCTS program, not in KataGo, right?

Choosing zero (i.e. the middle possible value in your Q range) can lead to problems for exactly the reason you described.

There isn't a standard way to do this, however. Some possible ways:
* Treat the Q value of an unexplored move as equal to the predicted value from the action-value head of the neural net.
    * Theoretically principled and could possibly be one of the best ways but uncommon in AlphaZero programs because most of them don't train an action-value head ("q value"), only a state value head. 
    * Even with such a head, one would probably want to also average these values with some weighting with things like the parent average value, due to the fact that there is correlation among the child value estimates, so that once some of the children are searched one can get a better estimate at runtime than the raw neural net action-value head by looking at the value of other children.
* Treat it as equal to a loss value, i.e. the lowest possible value. 
    * Original AlphaZero. Should work, but might not be the best - experimentation in Leela Zero a long time ago found some pathologies with this.
* Treat it as equal to the parent average value minus a tunable constant.
    * Leela Zero and (I think) Leela Chess Zero. Performs quite well once the constant is tuned, if the neural net is strong. Note that the root could possibly want a different tuned value than the rest of the tree (and the root being special is theoretically justifiable here, because at the root one is optimizing for simple regret whereas deeper in the tree one is optimizing for something closer to cumulative regret over playouts).
    * KataGo used this originally.
* Treat it as equal to the *equal-weighted* average value of all child nodes explored so far, minus a tunable constant.
    * Could also be tried, might have have also been used by some bots or MCTS implementations. Compared to using the parent average valu, this one will decay faster towards low values if you start to get bad values for each new child node explored, suppressing further exploration, while not doing so if new child nodes explored are still pretty good in value.
* Treat it as equal to parent average value * frac + parent raw neural net value * (1-frac) - c, where frac = (fraction of total policy mass that has at least one visit)^alpha, and where alpha and c are a tunable or adjustable constants.
    * KataGo's current method. In practice, behaves very similarly to Leela Zero's method. The way that it differs is that until most of the policy is explored, it will "trust" the raw neural net evaluation a bit more. For example, if the raw neural net evaluation thinks the position is very good, but the first move explored turns out to be a blunder, the parent average value will be dragged down heavily by the blunder value of the first move, so using parent average value will suppress exploration of other alternatives, whereas this method will give a high weight to explore alternatives to try to see if the parent's high raw neural net value was actually correct and it's just that the first move was wrong. But if more moves are tried and are still getting bad values, then it will drop back to the parent average value minus a constant.

**Comment by lightvector** [comment]
Note that if you are going to test more than one way, it's critical to optimize c_puct separately for each way before you compare their relative strength, and in the case of methods that have tunable constants, you will want to *jointly* optimize the tunable constant together with c_puct.

**Comment by Garbage123King** [comment]
Thank you very much!

**Comment by Garbage123King** [comment]
I'm confused about another detail:

In the KataGo paper, the formula for PUCT is given as:

PUCT(c) = V(c)+c_PUCT * P(c) * sum(N(c')) / (1 +N(c))

When a node selects a child for the first time, all the children have the same V(c), whether it's initialized to 0 or derived from the parent.

Also, sum(N(c')) = 0 because no child has been visited yet.

This means all children have the same PUCT(c) value, so on the first selection, does the node just choose a child uniformly at random? ( I think it can't be true)

@lightvector
